firedrake/functionspaceimpl.py

```diff
@@ -0,0 +1,736 @@
+        """Return the :class:`pyop2.Map` from exterior facets to
+        function space nodes. If present, bcs must be a tuple of
+        :class:`.DirichletBC`\s. In this case, the facet_node_map will return
+        negative node indices where boundary conditions should be
+        applied. Where a PETSc matrix is employed, this will cause the
+        corresponding values to be discarded during matrix assembly."""
+        # FIXME: these want caching of sorts
+        bc_list = [[] for _ in self]
+        if bcs:
+            for bc in bcs:
+                bc_list[bc.function_space().index].append(bc)
+        return op2.MixedMap(s.exterior_facet_node_map(bc_list[i])
+                            for i, s in enumerate(self._spaces))
+
+    @utils.cached_property
+    def exterior_facet_boundary_node_map(self):
+        '''The :class:`pyop2.MixedMap` from exterior facets to the nodes on
+        those facets. Note that this differs from
+        :meth:`exterior_facet_node_map` in that only surface nodes
+        are referenced, not all nodes in cells touching the surface.'''
+        return op2.MixedMap(s.exterior_facet_boundary_node_map for s in self._spaces)
+
+    def make_dat(self, val=None, valuetype=None, name=None, uid=None):
+        """Return a newly allocated :class:`pyop2.MixedDat` defined on the
+        :attr:`dof_dset` of this :class:`MixedFunctionSpace`."""
+        if val is not None:
+            assert len(val) == len(self)
+        else:
+            val = [None for _ in self]
+        return op2.MixedDat(s.make_dat(v, valuetype, "%s[cmpt-%d]" % (name, i), utils._new_uid())
+                            for i, (s, v) in enumerate(zip(self._spaces, val)))
+
+    @utils.cached_property
+    def _dm(self):
+        """A PETSc DM describing the data layout for fieldsplit solvers."""
+        dm = PETSc.DMShell().create()
+        dm.setAttr('__fs__', weakref.ref(self))
+        dm.setCreateFieldDecomposition(self.create_field_decomp)
+        dm.setCreateSubDM(self.create_subdm)
+        # FIXME: The Vec could be shared between functionspaces with
+        # the same dof_dset.
+        with self.make_dat().vec_ro as v:
+            dm.setGlobalVector(v.duplicate())
+        return dm
+
+    @utils.cached_property
+    def _ises(self):
+        return self.dof_dset.field_ises
+
+    @classmethod
+    def create_subdm(cls, dm, fields, *args, **kwargs):
+        W = dm.getAttr('__fs__')()
+        if len(fields) == 1:
+            # Subspace is just a single FunctionSpace.
+            subspace = W[fields[0]]
+        else:
+            # Need to build an MFS for the subspace
+            subspace = MixedFunctionSpace([W[f] for f in fields])
+        # Sub-DM is just the DM belonging to the subspace.
+        subdm = subspace._dm
+        # Keep hold of strong reference to created subspace (given we
+        # only hold a weakref in the shell DM)
+        W._subspaces.append(subspace)
```

Comment:
Return same subspace if asked for this again?  Now that `MFS` is not object cached we probably should.